### ISSUE TITLE
Ubuntu 18.04.02 Broken Since Package Upgrade (coredns "Pod sandbox changed, it will be killed...")

### ISSUE THREAD

**Issue by zaneclaes** [ubuntu/microk8s#506]
I'm breaking this out as a separate issue from my comments [here](https://github.com/ubuntu/microk8s/issues/48) @ktsakalozos , where I described that all my deployments stopped working on a machine that had been untouched for ~10 days simply because I did an `apt-get upgrade` and rebooted. No other changes were made.

The following steps reproduce the issue, which I've tried on both stable and edge channels. You can find the output of `microk8s.inspect` [here](https://github.com/ubuntu/microk8s/files/3289009/inspection-report-20190613_211044.tar.gz)

Each time, I ran just the following commands:
```
microk8s.reset
snap remove microk8s
sudo snap install microk8s --edge --classic
microk8s.enable dns storage
```
Between the uninstall & reinstall, I have also tried verifying that `/snap/microk8s` and `/var/snap/microk8s` and `cat /proc/mounts | grep microk8` are all empty. I have also tried rebooting at that time and double-checking that there are no updates via `snap refresh microk8s`. If I run `microk8s.kubectl get pods --all-namespaces -o wide` I see:
```
  IP           NODE    NOMINATED NODE   READINESS GATES
kube-system   coredns-5874dcd95f-sn2gh                1/1     Running   11         2m49s   10.1.1.144   house   <none>           <none>
kube-system   hostpath-provisioner-6d744c4f7c-pkhqj   1/1     Running   5          2m42s   10.1.1.143   house   <none>           <none>
```
Describing one of the pods, I see:
```
Name:               coredns-5874dcd95f-sn2gh
Namespace:          kube-system
Priority:           2000000000
PriorityClassName:  system-cluster-critical
Node:               house/192.168.1.100
Start Time:         Thu, 13 Jun 2019 20:59:10 -0700
Labels:             k8s-app=kube-dns
                    pod-template-hash=5874dcd95f
Annotations:        scheduler.alpha.kubernetes.io/critical-pod: 
Status:             Running
IP:                 10.1.1.149
Controlled By:      ReplicaSet/coredns-5874dcd95f
Containers:
  coredns:
    Container ID:  containerd://ec8410c66229805b25abc5560d3d30d02f39e63dd64ce5e01249e46582a52c60
    Image:         coredns/coredns:1.5.0
    Image ID:      sha256:7987f0908caf347ac98bc23ef590eaf659dcdca661d57fdada563ccfe7edb043
    Ports:         53/UDP, 53/TCP, 9153/TCP
    Host Ports:    0/UDP, 0/TCP, 0/TCP
    Args:
      -conf
      /etc/coredns/Corefile
    State:          Running
      Started:      Thu, 13 Jun 2019 21:02:30 -0700
    Last State:     Terminated
      Reason:       Unknown
      Exit Code:    255
      Started:      Thu, 13 Jun 2019 21:02:18 -0700
      Finished:     Thu, 13 Jun 2019 21:02:28 -0700
    Ready:          False
    Restart Count:  13
    Limits:
      memory:  170Mi
    Requests:
      cpu:        100m
      memory:     70Mi
    Liveness:     http-get http://:8080/health delay=60s timeout=5s period=10s #success=1 #failure=5
    Readiness:    http-get http://:8181/ready delay=0s timeout=1s period=10s #success=1 #failure=3
    Environment:  <none>
    Mounts:
      /etc/coredns from config-volume (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from coredns-token-t5bgj (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  config-volume:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      coredns
    Optional:  false
  coredns-token-t5bgj:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  coredns-token-t5bgj
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     CriticalAddonsOnly
                 node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason                  Age                    From               Message
  ----     ------                  ----                   ----               -------
  Normal   Scheduled               3m22s                  default-scheduler  Successfully assigned kube-system/coredns-5874dcd95f-sn2gh to house
  Normal   Pulling                 3m19s                  kubelet, house     Pulling image "coredns/coredns:1.5.0"
  Normal   Pulled                  3m15s                  kubelet, house     Container image "coredns/coredns:1.5.0" already present on machine
  Normal   Created                 3m14s                  kubelet, house     Created container coredns
  Normal   Started                 3m14s                  kubelet, house     Started container coredns
  Normal   SandboxChanged          3m4s                   kubelet, house     Pod sandbox changed, it will be killed and re-created.
  Normal   Pulled                  3m3s                   kubelet, house     Container image "coredns/coredns:1.5.0" already present on machine
  Normal   Created                 3m3s                   kubelet, house     Created container coredns
  Normal   Started                 3m3s                   kubelet, house     Started container coredns
  Normal   SandboxChanged          2m58s (x2 over 2m58s)  kubelet, house     Pod sandbox changed, it will be killed and re-created.
  Normal   Pulled                  2m57s                  kubelet, house     Container image "coredns/coredns:1.5.0" already present on machine
  Normal   Created                 2m57s                  kubelet, house     Created container coredns
  Normal   Started                 2m57s                  kubelet, house     Started container coredns
  Normal   SandboxChanged          2m39s (x2 over 2m40s)  kubelet, house     Pod sandbox changed, it will be killed and re-created.
  Normal   Pulled                  2m39s                  kubelet, house     Container image "coredns/coredns:1.5.0" already present on machine
  Normal   Created                 2m39s                  kubelet, house     Created container coredns
  Warning  Failed                  2m39s                  kubelet, house     Error: transport is closing
  Normal   SandboxChanged          2m33s (x2 over 2m34s)  kubelet, house     Pod sandbox changed, it will be killed and re-created.
  Normal   Pulled                  2m33s                  kubelet, house     Container image "coredns/coredns:1.5.0" already present on machine
  Normal   Created                 2m32s                  kubelet, house     Created container coredns
  Normal   Started                 2m32s                  kubelet, house     Started container coredns
  Normal   SandboxChanged          2m13s (x2 over 2m13s)  kubelet, house     Pod sandbox changed, it will be killed and re-created.
  Normal   Pulled                  2m12s                  kubelet, house     Container image "coredns/coredns:1.5.0" already present on machine
  Warning  Failed                  2m11s                  kubelet, house     Error: transport is closing
  Normal   SandboxChanged          2m5s (x2 over 2m6s)    kubelet, house     Pod sandbox changed, it will be killed and re-created.
  Normal   Pulled                  2m5s                   kubelet, house     Container image "coredns/coredns:1.5.0" already present on machine
  Normal   Created                 2m4s                   kubelet, house     Created container coredns
  Normal   Started                 2m4s                   kubelet, house     Started container coredns
  Normal   SandboxChanged          107s (x2 over 107s)    kubelet, house     Pod sandbox changed, it will be killed and re-created.
  Warning  FailedCreatePodSandBox  106s                   kubelet, house     Failed create pod sandbox: rpc error: code = Unavailable desc = transport is closing
  Normal   SandboxChanged          101s                   kubelet, house     Pod sandbox changed, it will be killed and re-created.
  Normal   Pulled                  101s                   kubelet, house     Container image "coredns/coredns:1.5.0" already present on machine
  Normal   Started                 100s                   kubelet, house     Started container coredns
  Normal   Created                 100s                   kubelet, house     Created container coredns
  Normal   SandboxChanged          88s (x2 over 89s)      kubelet, house     Pod sandbox changed, it will be killed and re-created.
  Normal   Pulled                  88s                    kubelet, house     Container image "coredns/coredns:1.5.0" already present on machine
  Normal   Created                 88s                    kubelet, house     Created container coredns
  Normal   Started                 87s                    kubelet, house     Started container coredns
  Normal   SandboxChanged          70s (x2 over 71s)      kubelet, house     Pod sandbox changed, it will be killed and re-created.
  Normal   Pulled                  69s                    kubelet, house     Container image "coredns/coredns:1.5.0" already present on machine
  Normal   Created                 69s                    kubelet, house     Created container coredns
  Warning  Failed                  69s                    kubelet, house     Error: transport is closing
  Normal   SandboxChanged          64s (x2 over 64s)      kubelet, house     Pod sandbox changed, it will be killed and re-created.
  Normal   Pulled                  63s                    kubelet, house     Container image "coredns/coredns:1.5.0" already present on machine
  Normal   Created                 63s                    kubelet, house     Created container coredns
  Normal   Started                 63s                    kubelet, house     Started container coredns
  Normal   SandboxChanged          46s (x2 over 46s)      kubelet, house     Pod sandbox changed, it will be killed and re-created.
  Normal   Pulled                  45s                    kubelet, house     Container image "coredns/coredns:1.5.0" already present on machine
  Normal   Created                 45s                    kubelet, house     Created container coredns
  Warning  Failed                  45s                    kubelet, house     Error: transport is closing
  Normal   SandboxChanged          39s (x2 over 40s)      kubelet, house     Pod sandbox changed, it will be killed and re-created.
  Normal   Pulled                  38s                    kubelet, house     Container image "coredns/coredns:1.5.0" already present on machine
  Normal   Created                 38s                    kubelet, house     Created container coredns
  Normal   Started                 37s                    kubelet, house     Started container coredns
  Normal   SandboxChanged          21s (x2 over 22s)      kubelet, house     Pod sandbox changed, it will be killed and re-created.
  Warning  FailedCreatePodSandBox  20s                    kubelet, house     Failed create pod sandbox: rpc error: code = Unavailable desc = transport is closing
  Normal   SandboxChanged          15s                    kubelet, house     Pod sandbox changed, it will be killed and re-created.
  Normal   Pulled                  15s                    kubelet, house     Container image "coredns/coredns:1.5.0" already present on machine
  Normal   Created                 15s                    kubelet, house     Created container coredns
  Normal   Started                 14s                    kubelet, house     Started container coredns
  Normal   SandboxChanged          3s (x2 over 3s)        kubelet, house     Pod sandbox changed, it will be killed and re-created.
  Normal   Pulled                  2s                     kubelet, house     Container image "coredns/coredns:1.5.0" already present on machine
  Normal   Created                 2s                     kubelet, house     Created container coredns
  Normal   Started                 2s                     kubelet, house     Started container coredns
```

**Comment by ktsakalozos** [comment]
We have a service that detects changes in the network so as to reconfigure the services accordingly. Can you try to stop that service and see if the service restarts stop? Try:
```
sudo systemctl stop snap.microk8s.daemon-apiserver-kicker.service
```

**Comment by ktsakalozos** [comment]
@zaneclaes could you also apt-get remove docker.io and containerd.io just to check we are not hitting any containerds clashes?

**Comment by zaneclaes** [comment]
tl;dr: it worked, thanks!

I suspected this was something with the container agent based on other threads, and also because I noticed something _very_ odd just before responding here. One of my k8s containers is `home-assistant` (IoT home automation). I woke up this morning and noticed all of my home automations (motion sensor lights, etc.) were all still working despite never applying the deployment after reinstalling microk8s.

@ktsakalozos I ran the first command, as well as: `sudo apt-get remove docker-ce docker-ce-cli containerd.io`

The DNS then successfully started after a disable/enable, and I was able to spin up the deployments.

**Comment by ktsakalozos** [comment]
Good to hear. Make sure you do not follow the edge channel as it gets frequent updates.

Out of personal interest, how do you deploy home-assistant into MicroK8s? Is there a guide I could follow? I would love to have it as an addon (`microk8s.enable home-assistant`)!

**Comment by zaneclaes** [comment]
There is a helm chart available, but I ended up hand-tuning my deployment instead. This is because I run multiple different servers off my home Linux machine behind an open-source front proxy based upon Envoy (still a WIP, but check it out at [Switchboard](https://github.com/zaneclaes/switchboard)).

Here's my Kubernetes `deployment.yml` for `home-assistant`. Note that I manually upgrade the version of the container as I see fit. Also note that the `/dev` directory is mounted along with a `securityContext` of `privileged` in order to allow access to the USB dongles used for `Z-Wave` and `ZHA`.

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: home-assistant-ingress
spec:
  backend:
    serviceName: home-assistant
    servicePort: 8111
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: home-assistant
  labels:
    app: home-assistant
spec:
  replicas: 1
  selector:
    matchLabels:
      app: home-assistant
  template:
    metadata:
      labels:
        app: home-assistant
    spec:
      volumes:
      - name: config
        hostPath:
          path: /home/zaneclaes/home-assistant/config
          type: Directory
      - name: dev
        hostPath:
          path: /dev
          type: Directory
      hostNetwork: true
      containers:
        - name: home-assistant
          image: "homeassistant/home-assistant:0.94.3"
          ports:
            - name: api
              containerPort: 8111
              protocol: TCP
          livenessProbe:
            httpGet:
              path: /
              port: api
            initialDelaySeconds: 30
            failureThreshold: 5
            timeoutSeconds: 5
          readinessProbe:
            httpGet:
              path: /
              port: api
            initialDelaySeconds: 30
            failureThreshold: 5
            timeoutSeconds: 5
          volumeMounts:
          - mountPath: /config
            name: config
          - mountPath: /dev
            name: dev
          securityContext:
            privileged: true
```

**Comment by zaneclaes** [comment]
@ktsakalozos the issue is back. Again, I rebooted the computer. Again, basic features are broken. Again, I tried reinstalling `microk8s`. I have confirmed that the docker container daemons did not get somehow re-added. When I inspect the `dns` pod, here's what I get:

```
  Normal   Created         40s                kubelet, house     Created container sidecar
  Normal   Created         40s                kubelet, house     Created container kubedns
  Normal   Pulled          40s                kubelet, house     Container image "gcr.io/google_containers/k8s-dns-kube-dns-amd64:1.14.7" already present on machine
  Normal   Pulled          40s                kubelet, house     Container image "gcr.io/google_containers/k8s-dns-dnsmasq-nanny-amd64:1.14.7" already present on machine
  Normal   Created         40s                kubelet, house     Created container dnsmasq
  Normal   Started         40s                kubelet, house     Started container dnsmasq
  Normal   Pulled          40s                kubelet, house     Container image "gcr.io/google_containers/k8s-dns-sidecar-amd64:1.14.7" already present on machine
  Normal   Started         40s                kubelet, house     Started container kubedns
  Normal   Started         39s                kubelet, house     Started container sidecar
  Normal   SandboxChanged  23s (x2 over 23s)  kubelet, house     Pod sandbox changed, it will be killed and re-created.
  Warning  InspectFailed   22s                kubelet, house     Failed to inspect image "gcr.io/google_containers/k8s-dns-sidecar-amd64:1.14.7": rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial unix /var/snap/microk8s/common/run/containerd.sock: connect: no such file or directory"
  Normal   Pulled          22s                kubelet, house     Container image "gcr.io/google_containers/k8s-dns-kube-dns-amd64:1.14.7" already present on machine
  Warning  InspectFailed   22s                kubelet, house     Failed to inspect image "gcr.io/google_containers/k8s-dns-dnsmasq-nanny-amd64:1.14.7": rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial unix /var/snap/microk8s/common/run/containerd.sock: connect: no such file or directory"
  Warning  Failed          22s                kubelet, house     Error: ImageInspectError
  Warning  Failed          22s                kubelet, house     Error: ImageInspectError
  Warning  Failed          22s                kubelet, house     Error: transport is closing
  Normal   SandboxChanged  17s (x2 over 17s)  kubelet, house     Pod sandbox changed, it will be killed and re-created
```

It seems to suggest that it cannot talk to the microk8s containerd, but again `microk8s.inspect` reports that everything is healthy. Here's the output:
[inspection-report-20190617_230848.tar.gz](https://github.com/ubuntu/microk8s/files/3299990/inspection-report-20190617_230848.tar.gz)

**Comment by zaneclaes** [comment]
Update: I re-read this thread and ran the command to stop the kicker service, and additionally disabled it so that it would not start on boot. I can now reboot the machine successfully, but don't understand why this was necessary.

Browsing through the log files in my last comment, I notice that the `apiserver` indeed seems to be detecting network changes as it is spammed with the following:
```
Jun 17 23:08:35 house microk8s.daemon-apiserver[7397]: I0617 23:08:35.259935    7397 asm_amd64.s:1337] ccResolverWrapper: sending new addresses to cc: [{etcd.socket:2379 0  <nil>}]
Jun 17 23:08:35 house microk8s.daemon-apiserver[7397]: I0617 23:08:35.259961    7397 asm_amd64.s:1337] balancerWrapper: got update addr from Notify: [{etcd.socket:2379 <nil>}]
Jun 17 23:08:35 house microk8s.daemon-apiserver[7397]: I0617 23:08:35.260083    7397 asm_amd64.s:1337] balancerWrapper: got update addr from Notify: [{etcd.socket:2379 <nil>}]
```

There also appears to be some suspicious spam in the `journal.log` for `containerd`:
```
Jun 17 23:08:34 house microk8s.daemon-containerd[6390]: time="2019-06-17T23:08:34-07:00" level=fatal msg="containerd-shim: ttrpc server failure" error="ttrpc: server closed"
Jun 17 23:08:34 house microk8s.daemon-containerd[6390]: time="2019-06-17T23:08:34-07:00" level=fatal msg="containerd-shim: ttrpc server failure" error="ttrpc: server closed"
Jun 17 23:08:34 house microk8s.daemon-containerd[6390]: time="2019-06-17T23:08:34-07:00" level=fatal msg="containerd-shim: ttrpc server failure" error="ttrpc: server closed"
```

**Comment by zaneclaes** [comment]
@ktsakalozos I discovered that my microk8s cluster was busted again this morning. It looks like the kicker service started itself again, despite having been `disable`d. It looks like `microk8s.start` re-enables the service, thus breaking itself on every startup. Even after disabling the service, a new error began appearing when creating some pods (but not others):
```
Failed create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "b80f79e8141773c16fc12e069dd418ba61624b5330029db661ab5017200df6fe": failed to allocate for range 0: no IP addresses available in range set: 10.1.1.1-10.1.1.254
```

I can't claim to know much about the internals of networking here, but clearly something is pretty messed up. Can you please provide guidance?
